### PR TITLE
Convert API to use the C interoperability features

### PR DESCRIPTION
**Pull request by dschwen**
Refs #103

This PR together with an app level change like this:

```
-      char cPhaseName[25];
-      int lcPhaseName = sizeof(cPhaseName);
-      ConvertToFortran(cPhaseName, lcPhaseName, _sp_phase_name[i].c_str());
-      char cSpeciesName[25];
-      int lcSpeciesName = sizeof(cSpeciesName);
-      ConvertToFortran(cSpeciesName, lcSpeciesName, _sp_species_name[i].c_str());
-      FORTRAN_CALL(Thermochimica::getoutputmolspeciesphase)
-      (cPhaseName,
-       &lcPhaseName,
-       cSpeciesName,
-       &lcSpeciesName,
-       &d._species_fractions[i],
-       &idbg,
-       30,
-       30);
+      Thermochimica::getOutputMolSpeciesPhase(_sp_phase_name[i].c_str(),
+                                              _sp_phase_name[i].length(),
+                                              _sp_species_name[i].c_str(),
+                                              _sp_species_name[i].length(),
+                                              &d._species_fractions[i],
+                                              &idbg);
```

seems to do the trick.

### CODE REVIEW

File: src/postprocess/GetOutputMolSpeciesPhase.f90 [hunk 1]
```diff
@@ -39,25 +39,29 @@
...
     character(30)                :: cTempPhase, cTempSpecies, cSearchPhase, cSearchSpecies
 
-    cSearchPhase = cPhase!TRIM(cPhase(1:min(30,lcPhase)))
-    cSearchSpecies = cSpecies!TRIM(cSpecies(1:min(30,lcSpecies)))
+    call c_f_pointer( cptr=c_loc(cSpecies), fptr=fSpecies )
+    call c_f_pointer( cptr=c_loc(cPhase), fptr=fPhase )
 
-    cSearchPhase = TRIM(cSearchPhase(1:min(30,lcPhase)))
-    cSearchSpecies = TRIM(cSearchSpecies(1:min(30,lcSpecies)))
+    cSearchPhase = TRIM(fPhase(1:min(30,lcPhase)))
```
Comment (dschwen) [inline src/postprocess/GetOutputMolSpeciesPhase.f90:63]: I suppose there is no need to trim anymore. I'll iterate on this.

**Comment by dschwen** [comment]
ping @maxposchmann

**Comment by dschwen** [comment]
Is CI enabled for this repo? I'm only seeing _Expected — Waiting for status to be reported_

**Comment by maxposchmann** [comment]
Tests are automated but run locally, so we've disabled them on external pull requests for security. I can run them manually as required. They aren't going to test these APIs anyway though.

**Comment by dschwen** [comment]
> external pull requests for security

You can totally trust me ...

**Comment by maxposchmann** [comment]
I've approved all the ones in queue now.

Thank you for working on this. Please let me know when you're done. I'd like to make matching changes in `Thermochimica-c.c` and run some tests before merging. FYI that file is our wrapper for all C applications we've worked on since BISON. Maybe you'd be better to judge than me: is it better (and easier to maintain) if the MOOSE tools you're working on also use this wrapper?

**Comment by dschwen** [comment]
I suggest a few changes to `Thermochimica-c.c` 

* make it `Thermochimica-c.C`
* use `std::string` (and do the `.c_str()`, `.length()` stuff in that wrapper
* put it all into a namespace

**Comment by dschwen** [comment]
There are some functions (like `Thermochimica`) which are called from both the C++ code as well as internally. We cannot make those `bind(C)` without putting them in a fortran _module_ (otherwise the linker expects fortran mangling and doesn't find the function).

**Comment by dschwen** [comment]
Same issue in the failing test 56 `TestThermo56.F90:(.text+0x1da): undefined reference to `getoutputchempot_'`. That function is now `bind(C)` and cannot be used from Fortran without putting it in a module.

**Comment by dschwen** [comment]
... or adding a thin `bind(C)` wrapper that calls the original fortran binding version...

Yeah, that's the least invasive option. Also in absence(?) of namespacing ion F90, I suggest a common prefix `ThermochimicaAPI_` for all  `bind(C)` API functions. If we create a C++ wrapper for all of those anyways the user will never see those names.

**Comment by maxposchmann** [comment]
Great, that sounds good to me. Maybe we can simplify the Fortran binding versions at the same time, since they won't need the string lengths for anything anymore.

**Comment by maxposchmann** [comment]
If it's alright with you, I'd like to merge this into a branch other than master and add you as a collaborator, so that it's easier for me to contribute to it in parallel. This would also solve the testing issue.

**Comment by dschwen** [comment]
@maxposchmann check out the latest version of this PR. Basic ideas:

1. All fortran functions that are to be called from C/C++ are marked `bind(C)` and their exported symbol names are prefixed with `TCAPI_` 
2. Fortran functions that are to be called from _both_ fortran and C/C++ are wrapped with a small `bind(C)` fortran function
3. I now distinguish between a C and a (new) C++ API, the C API should be backwards compatible
4. The new C++ API is namespaced, uses C++ strings, and pair/tuple return types rather than pointer arguments.

**Comment by maxposchmann** [comment]
Great, this is looking excellent.

We will need to create thin `bind(C)` wrappers for *all* API functions though, not just those called within Thermochimica. We support coupling with a few Fortran applications, and will need our APIs to continue working for those. I think the right way to handle this is to not make any of the existing APIs `bind(C)`, and to create a `[function]API` for each, as you did in `ThermochimicaAPI` and a few others. 

It might be cleanest to place all of these in a new file. Some other reorganization might also be in order.

I don't mind working on these changes. Did you give any thought to my suggestion about merging into a new branch? If you have an alternative method to preserve your history but also let me work on the PR, I'm open to suggestions.

**Comment by dschwen** [comment]
Merging into a new branch is fine with me! That way I could already update the bison submodule and continue with the Bison integration.

**Comment by maxposchmann** [comment]
Great, I've done that, next I'll branch off it to make my changes. When I'm ready, I'll ask you to review those. Finally, I have a couple of other groups I'd like to get review from before we merge into master.

**Comment by dschwen** [comment]
@maxposchmann if I may make a more general comment on the API; there are too many string operations involved in each API call in my opinion. The API could be  simplified if instead of passing _strings_ for phases, elements, and species, the user were to pass in _integer IDs_. 

1. it would eliminate many `TRIM` and `ADJUSTL` operations
2. it would eliminate many string comparisons
3. it would eliminate many boiler plate loops over element/phase/species name lists 
4. it would simplify function signatures (getting rid of string+string length parameters in favor of more interoperable integer parameters)

**Comment by maxposchmann** [comment]
@dschwen I completely agree. Unfortunately, such changes are not possible. Due to Thermochimica's manipulation of the database, the user or programmer on the other side of the API cannot know which indices to request, and in fact they might change between cells or timesteps in an application like MOOSE if the elements present change.

The most we could do in this direction is create functions to look up the correct indices before they are needed. I think this would only address point 4 on your list, while adding some new complexity in call sequencing. I suppose we could package these functions together?

If I had infinite time, I would undertake a review of which `TRIM` and `ADJUSTL` operations are still necessary. I suspect the answer is already very few, as I have tried to manage strings more uniformly where they are read.

**Pull request merged**